integration-tests/src/test/java/com/datastax/oss/driver/api/core/metadata/NodeStateIT.java

```diff
@@ -0,0 +1,369 @@
+
+  public @Rule SimulacronRule simulacron = new SimulacronRule(ClusterSpec.builder().withNodes(2));
+
+  public @Rule ClusterRule cluster =
+      ClusterRule.builder(simulacron)
+          .withOptions(
+              "connection.pool.local.size = 2",
+              "connection.reconnection-policy.max-delay = 1 second")
+          .build();
+
+  private InternalDriverContext driverContext;
+  private final BlockingQueue<NodeStateEvent> stateEvents = new LinkedBlockingDeque<>();
+
+  private BoundNode simulacronControlNode;
+  private BoundNode simulacronRegularNode;
+  private DefaultNode metadataControlNode;
+  private DefaultNode metadataRegularNode;
+
+  @Before
+  public void setup() {
+    driverContext = (InternalDriverContext) cluster.cluster().getContext();
+    driverContext.eventBus().register(NodeStateEvent.class, stateEvents::add);
+
+    // Sanity check: the driver should have connected to simulacron
+    ConditionChecker.checkThat(
+            () ->
+                // 1 control connection + 2 pooled connections per node
+                simulacron.cluster().getActiveConnections() == 5)
+        .as("Connections established")
+        .before(10, TimeUnit.SECONDS)
+        .becomesTrue();
+
+    // Find out which node is the control node, and identify the corresponding Simulacron and driver
+    // metadata objects.
+    simulacronControlNode = simulacronRegularNode = null;
+    for (BoundNode boundNode : simulacron.cluster().getNodes()) {
+      if (boundNode.getActiveConnections() == 3) {
+        simulacronControlNode = boundNode;
+      } else {
+        simulacronRegularNode = boundNode;
+      }
+    }
+    assertThat(simulacronControlNode).isNotNull();
+    assertThat(simulacronRegularNode).isNotNull();
+
+    Map<InetSocketAddress, Node> nodesMetadata = cluster.cluster().getMetadata().getNodes();
+    metadataControlNode =
+        (DefaultNode) nodesMetadata.get(simulacronControlNode.inetSocketAddress());
+    metadataRegularNode =
+        (DefaultNode) nodesMetadata.get(simulacronRegularNode.inetSocketAddress());
+  }
+
+  @Test
+  public void should_report_connections_for_healthy_nodes() {
+    ConditionChecker.checkThat(
+            () -> {
+              assertThat(metadataControlNode).isUp().hasOpenConnections(3).isNotReconnecting();
+              assertThat(metadataRegularNode).isUp().hasOpenConnections(2).isNotReconnecting();
+            })
+        .as("Node metadata up-to-date")
+        .before(10, TimeUnit.SECONDS)
+        .becomesTrue();
+  }
+
+  @Test
+  public void should_keep_regular_node_up_when_still_one_connection() {
+    simulacronRegularNode.rejectConnections(0, RejectScope.UNBIND);
+    NodeConnectionReport report = simulacronRegularNode.getConnections();
+    simulacron.cluster().closeConnection(report.getConnections().get(0), CloseType.DISCONNECT);
+
+    ConditionChecker.checkThat(
+            () -> assertThat(metadataRegularNode).isUp().hasOpenConnections(1).isReconnecting())
```

Comment:
Awesome, change to allow asserts in `ConditionChecker` looks really nice, code is much more concise and nice that we can use custom assertions.